crates/wasi-common/yanix/src/sys/linux/mod.rs

```diff
@@ -8,8 +8,23 @@ use crate::dir::SeekLoc;
 use std::io::Result;
 
 impl SeekLoc {
+    #[cfg(target_pointer_width = "64")]
     pub unsafe fn from_raw(loc: i64) -> Result<Self> {
         let loc = loc.into();
         Ok(Self(loc))
     }
+    #[cfg(target_pointer_width = "32")]
+    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
+        // The cookie (or `loc`) is an opaque value, and applications aren't supposed to do
+        // arithmetic on them or pick their own values or have any awareness of the numeric
+        // range of the values. They're just supposed to pass back in the values that we
+        // give them. And any value we give them will be convertable back to `long`,
+        // because that's the type the OS gives them to us in. So return an `EINVAL`.
+        use std::convert::TryInto;
+        use std::io::Error;
+        let loc = loc
+            .try_into()
+            .map_err(|_| Error::from_raw_os_error(libc::EINVAL))?;
+        Ok(Self(loc))
+    }
```

Comment:
And now that I think about it, we can do a similar thing here -- just have one version for both 32-bit and 64-bit platforms, which returns a `Result<libc::c_long>`, and we can rely on the compiler to optimize away the `map_err` in the case where `c_long` is 64-bit.